File: 03/main.js
const fs = require('node:fs');
fs.readFile('input.txt', 'utf8', (err, data) => {
    const lines = data.split("\n");

    let [x, y]   = [0, 0];  // Part 1
    let [sx, sy] = [0, 0];  // Part 2
    let [rx, ry] = [0, 0];  // Part 2

    const visited_p1 = new Set(["0,0"]);  // Part 1, SET Include starting position
    const visited_p2 = new Set(["0,0"]);  // Part 2, SET Include starting position

    for (const line of lines) {
        for (let i = 0; i < line.length; i++) {
            // Part 1
            if (line[i] === ">" ) x++;
            if (line[i] === "<" ) x--;
            if (line[i] === "^" ) y++;
            if (line[i] === "v" ) y--;
            if (! visited_p1.has(`${x},${y}`)) visited_p1.add(`${x},${y}`);
            
            // Part 2
            if (i % 2 == 0) {
                if (line[i] === ">" ) sx++;
                if (line[i] === "<" ) sx--;
                if (line[i] === "^" ) sy++;
                if (line[i] === "v" ) sy--;
                if (! visited_p2.has(`${sx},${sy}`)) visited_p2.add(`${sx},${sy}`);
            }
            else {
                if (line[i] === ">" ) rx++;
                if (line[i] === "<" ) rx--;
                if (line[i] === "^" ) ry++;
                if (line[i] === "v" ) ry--;
                if (! visited_p2.has(`${rx},${ry}`)) visited_p2.add(`${rx},${ry}`);
            }
        }
    }

    console.log(`Part1: ${visited_p1.size}`);
    console.log(`Part2: ${visited_p2.size}`);
});

File: 03/main.py
with open("input.txt") as file:
    lines = file.readlines()
    for line in lines:
        x, y = 0, 0
        sx, sy = 0, 0
        rx, ry = 0, 0
        visited_p1 = {(0, 0)}   # SET Include starting position
        visited_p2 = {(0, 0)}   # SET Include starting position

        for i, c in enumerate(line):
            # Part 1
            if c == ">": x += 1
            if c == "<": x -= 1
            if c == "^": y += 1
            if c == "v": y -= 1
            if (x, y) not in visited_p1: visited_p1.add((x, y))
            
            # Part 2 (Santa and RoboSanta each handle one instruction)
            if not i % 2:
                # Santa's moves
                if c == ">": sx += 1
                if c == "<": sx -= 1
                if c == "^": sy += 1
                if c == "v": sy -= 1
                if (sx, sy) not in visited_p2: visited_p2.add((sx, sy))
            else:
                # RoboSanta's moves
                if c == ">": rx += 1
                if c == "<": rx -= 1
                if c == "^": ry += 1
                if c == "v": ry -= 1
                if (rx, ry) not in visited_p2: visited_p2.add((rx, ry))

        print(f"Part1: {len(visited_p1)}")
        print(f"Part2: {len(visited_p2)}")